specs/_features/ePBS/builder.md

```diff
@@ -0,0 +1,229 @@
+
+
+## Introduction
+
+This document represents the expected behavior of an "honest builder" with respect to ePBS of the Ethereum proof-of-stake protocol. 
+This document does not distinguish between a "node" (i.e. the functionality of following and reading the beacon chain) and a "builder" (i.e. the functionality of actively participating in consensus).
+The separation of concerns between these (potentially) two pieces of software is left as a design decision that is out of scope.
+
+A builder is an entity that participates in the block building of the Ethereum proof-of-stake protocol. This is an optional role for users in which they can post ETH as collateral and build blocks to see financial returns in exchange for building the protocol.
+In this design, builder is an extension of validator using a different withdrawal prefix `0x0b`. The protocol onboards builder by upgrading validators into builders if minimal staked balance is reached.
+
+## Becoming a builder
+
+### Initialization
+
+A builder must initialize many parameters locally before submitting a deposit and joining the builder registry.
+
+Refer key general from the phase 0 validator spec.
+
+##### Builder withdrawal credential
+
+Builder Withdrawal credentials with the builder withdrawal prefix specify
+a 20-byte Eth1 address `eth1_withdrawal_address` as the recipient for all withdrawals.
+The `eth1_withdrawal_address` can be the address of either an externally owned account or of a contract.
+
+The `withdrawal_credentials` field must be such that:
+
+* `withdrawal_credentials[:1] == BUILDER_WITHDRAWAL_PREFIX` [New in ePBS]
+* `withdrawal_credentials[1:12] == b'\x00' * 11`
+* `withdrawal_credentials[12:] == eth1_withdrawal_address`
+
+### Submit deposit
+
+No change from phase 0 validator spec.
+
+### Process deposit
+
+No change from phase 0 validator spec.
+
+### Activation
+
+No change from phase 0 validator spec.
+
+
+## Beacon chain responsibilities
+
+
+### Signed execution payload header envelope proposal
+
+#### Constructing the `SignedExecutionPayloadHeaderEnvelope`
+
+First, the builder need obtain an execution payload. The builder building on top of block on top of `state` must ake the following actions through execution layer
+
+1. Set `payload_id = prepare_execution_payload(state, pow_chain, safe_block_hash, finalized_block_hash, suggested_fee_recipient, execution_engine)`, where:
+  * `state` is the state object after applying `process_slots(state, slot)` transition to the resulting state of the parent block processing
+  * `safe_block_hash` is the return value of the `get_safe_execution_payload_hash(store: Store)` function call
+  * `finalized_block_hash` is the hash of the latest finalized execution payload (`Hash32()` if none yet finalized)
+  * `suggested_fee_recipient` is the value suggested to be used for the `fee_recipient` field of the builder
+
+
+```python
+def prepare_execution_payload(state: BeaconState,
```

Comment:
The payload attributes should contain the inclusion list summary at the very least, these are the txs that the payload will need to satisfy. The EL should have already cached the transactions when the builder notified of the IL before.